lib/iris/fileformats/rules.py

```diff
@@ -345,6 +376,34 @@ def _resolve_factory_references(
             aux_factory = factory.factory_class(*args)
             cube.add_aux_factory(aux_factory)
 
+    # In the case of multiple references which vary on a new dimension
+    # (such as time-dependent orography or surface-pressure), the cube may get replaced
+    # by one with a new dimension.
+    # In that case we must update the factory so its dependencies are coords of the
+    # new cube.
+    cube_coord_ids = [
+        id(coord) for coord, _ in cube._dim_coords_and_dims + cube._aux_coords_and_dims
+    ]
+    for factory in cube.aux_factories:
+        for name, dep in list(factory.dependencies.items()):
+            if id(dep) not in cube_coord_ids:
+                factory.update(dep, cube.coord(dep))
+
+    return cube
+
+
+class MultipleReferenceFieldDetector(threading.local):
+    def __init__(self):
+        self.found_multiple_refs = False
+
+
+# A single global object (per thread) to record whether multiple reference fields
```

Comment:
If there is an instance per thread, is there any risk that behaviour is dependant on how multiprocessing is done? Or is the relevant loading done entirely within a single thread?

Reply:
I believe the loading code is all single-threaded in operation.  It has to be since, while it has the input file open, it might not be safe to run parallel loading tasks since the file system interface may not be thread-safe (e.g. as for netcdf4-python) .